<a href="https://colab.research.google.com/github/JoaoPedro-SA/API-de-Gerenciamento-Escolar/blob/main/ATV_3_kim_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



# Lista 3 Respostas (codigo)

In [2]:
import pandas as pd
from collections import Counter

# =========================
# 1. TEXTO DE TREINAMENTO
# =========================
texto = "estudo ads na faculdade particular estudo ads na impacta estudo engenharia na poli"

# Pré-processamento simples
palavras = texto.lower().split()

# =========================
# 2. CRIAR BIGRAMAS
# =========================
bigramas = []

for i in range(len(palavras) - 1):
    palavra_atual = palavras[i]
    proxima_palavra = palavras[i + 1]
    bigramas.append((palavra_atual, proxima_palavra))

# Criar DataFrame
df = pd.DataFrame(bigramas, columns=["palavra_atual", "proxima_palavra"])

print("DataFrame de Bigramas:")
print(df)
print("\n")

# =========================
# 3. FUNÇÃO DE PREVISÃO
# =========================
def prever_proxima_palavra(contexto):
    # Filtrar onde a palavra atual aparece
    filtro = df[df["palavra_atual"] == contexto]

    if filtro.empty:
        return f"Nenhuma ocorrência encontrada para '{contexto}'"

    # Contar frequência das próximas palavras
    contagem = Counter(filtro["proxima_palavra"])
    total = sum(contagem.values())

    # Calcular probabilidades
    probabilidades = {}
    for palavra, freq in contagem.items():
        prob = freq / total
        probabilidades[palavra] = prob

    # Ordenar por maior probabilidade
    probabilidades_ordenadas = dict(sorted(probabilidades.items(), key=lambda x: x[1], reverse=True))

    return probabilidades_ordenadas

# =========================
# 4. TESTES
# =========================
print("Teste com 'estudo':")
print(prever_proxima_palavra("estudo"))
print("\n")

print("Teste com 'engenharia':")
print(prever_proxima_palavra("engenharia"))
print("\n")

print("Teste com 'na':")
print(prever_proxima_palavra("na"))

DataFrame de Bigramas:
   palavra_atual proxima_palavra
0         estudo             ads
1            ads              na
2             na       faculdade
3      faculdade      particular
4     particular          estudo
5         estudo             ads
6            ads              na
7             na         impacta
8        impacta          estudo
9         estudo      engenharia
10    engenharia              na
11            na            poli


Teste com 'estudo':
{'ads': 0.6666666666666666, 'engenharia': 0.3333333333333333}


Teste com 'engenharia':
{'na': 1.0}


Teste com 'na':
{'faculdade': 0.3333333333333333, 'impacta': 0.3333333333333333, 'poli': 0.3333333333333333}


### 🔹 Questão para Reflexão (Storytelling)

**"O seu modelo previu 'poli' após a palavra 'na' em um contexto de engenharia. Se mudarmos o contexto para 'ADS', a probabilidade de 'poli' deve subir ou descer? Justifique utilizando o conceito de Probabilidade Condicional."**

**Resposta:**

Se mudarmos o contexto de 'engenharia' para 'ADS', a probabilidade de 'poli' após a palavra 'na' **deve descer (ou até mesmo ser zero, dependendo do corpus)**.

**Justificativa (Probabilidade Condicional):**

O modelo de bigramas calcula a probabilidade da próxima palavra *dada a palavra atual* (P(Próxima | Atual)).

No texto de treinamento:
`estudo ads na faculdade particular estudo ads na impacta estudo engenharia na poli`

Quando o contexto é 'engenharia', a sequência `'engenharia na poli'` aparece uma vez. Assim, $P('poli' | 'na', \text{depois de 'engenharia'})$ é alta (100% nesse corpus limitado, já que `'na'` só ocorre depois de `'engenharia'` quando é seguido por `'poli'` nesse contexto específico).

Agora, vamos considerar o contexto 'ADS':

As ocorrências de `'na'` após `'ads'` são:
- `'ads na faculdade'`
- `'ads na impacta'`

Nesses casos, a palavra que segue `'na'` nunca é 'poli'. Portanto, a probabilidade condicional de $P('poli' | 'na', \text{depois de 'ads'})$ seria 0% neste corpus específico.

Mesmo em um corpus maior onde 'poli' pudesse aparecer após 'na' em um contexto de 'ADS' (por exemplo, 'ads na poli' se referindo a uma localização física), a frequência de 'poli' após 'na' seria diluída pelas outras palavras que seguem 'na' em um contexto 'ADS', como 'faculdade' e 'impacta'. Isso faria com que a probabilidade de 'poli' descesse em comparação com o contexto de 'engenharia', onde `'na'` era imediatamente seguido por `'poli'` na única ocorrência relevante.

# Lista 3 Respostas (questões)


### 🔹 Parte 1 – Nível Básico

**1. Moeda**
Uma moeda honesta tem 2 resultados possíveis: {Cara, Coroa}
$P(Cara) = \frac{1}{2} = 0,5 = 50\%$

**2. Dado**
Números pares: {2, 4, 6} → 3 resultados de 6
$P(par) = \frac{3}{6} = \frac{1}{2} = 50\%$

**3. Urna**
Total de bolas: $5 + 3 + 2 = 10$
Bolas pretas: 3
$P(preta) = \frac{3}{10} = 0,3 = 30\%$

**4. Cartas**
Um baralho tem 52 cartas e 4 ases
$P(Ás) = \frac{4}{52} = \frac{1}{13} \approx 7,69\%$

**5. Dias da semana**
Dias: 7
Começam com “S”: Segunda e Sexta → 2
$P = \frac{2}{7} \approx 28,57\%$

**6. Probabilidade complementar**
$P(não chover) = 1 - 0,35 = 0,65 = 65\%$

### 🔹 Nível Intermediário

**1. Dois dados (soma = 7)**
Combinações favoráveis:
(1,6), (2,5), (3,4), (4,3), (5,2), (6,1) → 6 casos
Total: 36
$P = \frac{6}{36} = \frac{1}{6} \approx 16,67\%$

**2. Moeda duas vezes**
$P(cara \text{ e } cara) = \frac{1}{2} \times \frac{1}{2} = \frac{1}{4} = 25\%$

**3. Sem reposição**
$P = \frac{3}{10} \times \frac{2}{9} = \frac{6}{90} = \frac{1}{15} \approx 6,67\%$

**4. União de eventos**
Menor que 3: {1,2}
Maior que 5: {6}
Total favorável: 3
$P = \frac{3}{6} = \frac{1}{2} = 50\%$

**5. União (futebol ou basquete)**
$P(F \cup B) = P(F) + P(B) - P(F \cap B)$
$= 0,6 + 0,4 - 0,2 = 0,8 = 80\%$

**6. Probabilidade condicional**
Números pares: {2,4,6} → 3
$P(6 \mid par) = \frac{1}{3} \approx 33,33\%$


### 🔹 Desafios Intermediários

**1. Senha (3 dígitos iguais)**
Total de senhas: $10^3 = 1000$
Favoráveis: (000, 111, ..., 999) → 10
$P = \frac{10}{1000} = \frac{1}{100} = 1\%$

**2. Com reposição**
$P(\text{verde depois azul}) = \frac{4}{10} \times \frac{6}{10} = \frac{24}{100} = 24\%$

**3. Comissão (1 homem e 1 mulher)**
Total de formas de escolher 2 pessoas de 5:
$C(5,2) = 10$
Casos favoráveis:
$3 \times 2 = 6$
$P = \frac{6}{10} = 60\%$

### 🔹 Parte 2 – Exercícios Computacionais (explicação teórica)

**1. Monte Carlo (dados)**
Simula milhares de lançamentos e calcula:
$\text{frequência relativa} = \frac{\text{ocorrências}}{\text{total}}$

**2. Estimativa de π**
$\pi \approx 4 \times \frac{\text{pontos dentro do círculo}}{\text{total de pontos}}$

**3. Sorteio de nomes**
A probabilidade muda a cada rodada:
$P = \frac{1}{\text{quantidade restante}}$

**4. Monty Hall**
Trocar porta → ~66%
Não trocar → ~33%

**5. Random Walk**
Média tende a ficar próxima de 0 (equilíbrio entre +1 e -1)

### 🔹 Problema do Aniversário

**Pergunta A (23 pessoas)**
A probabilidade de coincidência é:
$> 50\% (\approx 50,7\%)$
👉 Ou seja: é maior que 50%

**Pergunta B**
Para 100% de certeza:
366 pessoas
👉 Pelo princípio da casa dos pombos.